Sources/Core/Microsoft.StreamProcessing/Aggregates/TopKState.cs

```diff
@@ -0,0 +1,236 @@
+        {
+            this.k = k;
+            this.rankComparer = rankComparer;
+            this.currentValues = new SortedMultiSet<T>(generator);
+            this.previousValues = new SortedMultiSet<T>(generator);
+        }
+
+        public void Add(T input, long timestamp)
+        {
+            if (timestamp > currentTimestamp)
+            {
+                MergeCurrentToPrevious();
+                currentTimestamp = timestamp;
+            }
+            else if (timestamp < currentTimestamp)
+            {
+                throw new ArgumentException("Invalid timestamp");
+            }
+
+            currentValues.Add(input);
+
+            var toRemove = currentValues.TotalCount - k;
+            if (toRemove > 0)
+            {
+                var min = currentValues.GetMinItem();
+                if (toRemove == min.Count)
+                {
+                    currentValues.Remove(min.Item, min.Count);
+                }
+                else if (toRemove > min.Count)
+                {
+                    throw new InvalidOperationException("CurrentValues has more items than required");
+                }
+            }
+        }
+
+        public void Remove(T input, long timestamp)
+        {
+            if (timestamp < currentTimestamp)
+            {
+                previousValues.Remove(input);
+            }
+            else if (timestamp == currentTimestamp)
+            {
+                currentValues.Remove(input);
+            }
+            else
+            {
+                throw new ArgumentException("Invalid timestamp");
+            }
+        }
+
+        public void RemoveAll(ITopKState<T> other)
+        {
+            if (other.Count != 0)
+            {
+                if (other is HoppingTopKState<T> otherTopK)
+                {
+                    if (otherTopK.currentTimestamp >= currentTimestamp)
+                    {
+                        throw new ArgumentException("Cannot remove entries with current or future timestamp");
+                    }
+                    previousValues.RemoveAll(otherTopK.currentValues);
+                    previousValues.RemoveAll(otherTopK.previousValues);
+                }
+                else
+                {
+                    throw new InvalidOperationException("Cannot remove non-HoppingTopKState from HoppingTopKState");
+                }
+            }
+        }
+
+        public SortedMultiSet<T> GetSortedValues()
+        {
+            if (previousValues.IsEmpty)
+                return currentValues;
+            else
+            {
+                MergeCurrentToPrevious();
+                return previousValues;
+            }
+        }
+
+        private void MergeCurrentToPrevious()
+        {
+            if (!currentValues.IsEmpty)
+            {
+                // Swap so we merge small onto larger
+                if (previousValues.UniqueCount < currentValues.UniqueCount)
```

Comment:
>if (previousValues.UniqueCount < currentValues.UniqueCount) [](start = 16, length = 59)

When would this happen? Doesn't seem likely or significant enough to warrant the optimization

Reply:
Oh i guess most often on the first Merge?

---
In reply to: [356294107](https://github.com/microsoft/Trill/pull/121#discussion_r356294107) [](ancestors = 356294107)